src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -1020,23 +1137,144 @@ impl Manager {
-
-            m.insert(key.to_string(), p);
+            m.insert(key.to_string(), format!("{}/{}", mnt.unwrap(), cpath));
         }
 
-        let cg = new_cgroup(cgroups::hierarchies::auto(), cpath)?;
+        Ok((m, mounts))
+    }
 
-        Ok(Self {
-            paths: m,
-            mounts,
-            // rels: paths,
-            cpath: cpath.to_string(),
-            cgroup: cg,
-        })
+    fn setup_whitelist_mode(cgroup: &cgroups::Cgroup) -> Result<()> {
+        // Insert two rules: `b *:* rwm` and `c *:* rwm`.
+        // The reason of not inserting `a *:* rwm` is that the Linux kernel
+        // will deny writing `a` to `devices.allow` once a cgroup has
+        // children. You can refer to
+        // https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt.
+        let res = cgroups::Resources {
+            devices: cgroups::DeviceResources {
+                devices: vec![
+                    DeviceResource {
+                        allow: true,
+                        devtype: DeviceType::Block,
+                        major: -1,
+                        minor: -1,
+                        access: vec![
+                            DevicePermissions::Read,
+                            DevicePermissions::Write,
+                            DevicePermissions::MkNod,
+                        ],
+                    },
+                    DeviceResource {
+                        allow: true,
+                        devtype: DeviceType::Char,
+                        major: -1,
+                        minor: -1,
+                        access: vec![
+                            DevicePermissions::Read,
+                            DevicePermissions::Write,
+                            DevicePermissions::MkNod,
+                        ],
+                    },
+                ],
+            },
+            ..Default::default()
+        };
+        cgroup.apply(&res)?;
+
+        Ok(())
+    }
+
+    fn setup_blacklist_mode(cgroup: &cgroups::Cgroup) -> Result<()> {
+        #[allow(unused_mut)]
+        let mut dev_res_list = vec![DeviceResource {
+            allow: false,
+            devtype: DeviceType::All,
+            major: -1,
+            minor: -1,
+            access: vec![
+                DevicePermissions::Read,
+                DevicePermissions::Write,
+                DevicePermissions::MkNod,
+            ],
+        }];
+        // Do not append default allowed devices for simplicity while
+        // testing.
+        #[cfg(not(test))]
+        dev_res_list.append(&mut default_allowed_devices());
+
+        let res = cgroups::Resources {
+            devices: cgroups::DeviceResources {
+                devices: dev_res_list,
+            },
+            ..Default::default()
+        };
+        cgroup.apply(&res)?;
+
+        Ok(())
+    }
+
+    /// Check if OCI spec contains a whiltlist rule.
+    ///
+    /// # Returns
+    ///
```

Comment:
So what's the whitelist and blacklist? Maybe the white list is: deny all first and then allow the specified devices one by one; and the black list is on the opposite side: allow all first, and then deny the specified devices one by one.

Reply:
Fixed. The previous `whitelist` is replaced with `allowed_all`. The `blacklist` is replaced with `whitelist`.